scripts/error-codes/minify-error-messages.js

```diff
@@ -0,0 +1,133 @@
+
+  return {
+    visitor: {
+      CallExpression(path, file) {
+        const node = path.node;
+        if (path.get('callee').isIdentifier({name: 'invariant'})) {
+          // Turns this code:
+          //
+          // invariant(condition, 'A %s message that contains %s', adj, noun);
+          //
+          // into this:
+          //
+          // if (!condition) {
+          //   if (__DEV__) {
+          //     throw ReactError(`A ${adj} message that contains ${noun}`);
+          //   } else {
+          //     throw ReactErrorProd(ERR_CODE, [adj, noun]);
+          //   }
+          // }
+          //
+          // where ERR_CODE is an error code: a unique identifier (a number
+          // string) that references a verbose error message. The mapping is
+          // stored in `scripts/error-codes/codes.json`.
+          const condition = node.arguments[0];
+          const errorMsgLiteral = evalToString(node.arguments[1]);
+          const errorMsgExpressions = Array.from(node.arguments.slice(2));
+          const errorMsgQuasis = errorMsgLiteral
+            .split('%s')
+            .map(raw => t.templateElement({raw, cooked: String.raw({raw})}));
+
+          // Import ReactError
+          const reactErrorIdentfier = file.addImport(
+            'shared/ReactError',
+            'default',
+            'ReactError'
+          );
+
+          // Outputs:
+          //   throw ReactError(`A ${adj} message that contains ${noun}`);
+          const devThrow = t.throwStatement(
+            t.callExpression(reactErrorIdentfier, [
+              t.templateLiteral(errorMsgQuasis, errorMsgExpressions),
+            ])
+          );
+
+          // Avoid caching because we write it as we go.
+          const existingErrorMap = JSON.parse(
+            fs.readFileSync(__dirname + '/codes.json', 'utf-8')
+          );
+          const errorMap = invertObject(existingErrorMap);
+
+          let prodErrorId = errorMap[errorMsgLiteral];
+          if (prodErrorId === undefined) {
+            // There is no error code for this message. We use a lint rule to
+            // enforce that messages can be minified, so assume this is
+            // intentional and exit gracefully.
+            //
+            // Outputs:
+            //   if (!condition) {
+            //     throw ReactError(`A ${adj} message that contains ${noun}`);
+            //   }
+            path.replaceWith(
+              t.ifStatement(
+                t.unaryExpression('!', condition),
+                t.blockStatement([devThrow])
+              )
+            );
+            return;
+          }
+          prodErrorId = parseInt(prodErrorId, 10);
+
+          // Import ReactErrorProd
+          const reactErrorProdIdentfier = file.addImport(
+            'shared/ReactErrorProd',
+            'default',
+            'ReactErrorProd'
+          );
+
+          // Outputs:
+          //   throw ReactErrorProd(ERR_CODE, [adj, noun]);
```

Comment:
Can we make it `throw ReactErrorProd(ERR_CODE, adj, noun);`? Then wouldn't need `[]` in minified output on every callsite.

Reply:
I wasn't sure about this part so I just asked Sebastian and his rationale was that using arguments will make ReactErrorProd slightly slower, but using an array will likely make all the functions that throw slightly slower to compile, so it's hard to say which way is better. But since ReactErrorProd is in an error path, and fewer bytes is generally better, no array is good.